File: go/fory/row/reader.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package row

import (
	"encoding/binary"
	"fmt"
	"math"
)

// RowReader provides O(1) random-access reads from a Standard Row Format byte slice.
// Fields can be read independently without deserializing the entire row.
type RowReader struct {
	buf       []byte
	numFields int
}

// NewRowReader wraps buf for reading a row with numFields fields.
// Panics if numFields is negative or buf is smaller than the fixed region.
// Per ADR-4: validate once at construction; individual read methods then assume validity.
func NewRowReader(buf []byte, numFields int) *RowReader {
	if numFields < 0 {
		panic("row: numFields must be non-negative")
	}
	if need := FixedRegionSize(numFields); len(buf) < need {
		panic(fmt.Sprintf("row: buffer too small: need %d bytes for %d fields, got %d",
			need, numFields, len(buf)))
	}
	return &RowReader{buf: buf, numFields: numFields}
}

// IsNull returns true if fieldIdx is null.
// Per spec (confirmed by Java BitUtils.isSet): bit=1 means null.
func (r *RowReader) IsNull(fieldIdx int) bool {
	return (r.buf[fieldIdx>>3]>>uint(fieldIdx&7))&1 == 1
}

// readPrimSlot reads the full 8-byte slot for fieldIdx as a uint64.
func (r *RowReader) readPrimSlot(fieldIdx int) uint64 {
	return binary.LittleEndian.Uint64(r.buf[SlotOffset(fieldIdx, r.numFields):])
}

// ReadBool reads the bool stored in fieldIdx's slot (0x00 = false, any other = true).
func (r *RowReader) ReadBool(fieldIdx int) bool {
	return r.buf[SlotOffset(fieldIdx, r.numFields)] != 0
}

// ReadInt8 reads the int8 stored in fieldIdx's slot.
func (r *RowReader) ReadInt8(fieldIdx int) int8 {
	return int8(r.buf[SlotOffset(fieldIdx, r.numFields)])
}

// ReadInt16 reads the int16 stored in fieldIdx's slot (little-endian).
func (r *RowReader) ReadInt16(fieldIdx int) int16 {
	return int16(binary.LittleEndian.Uint16(r.buf[SlotOffset(fieldIdx, r.numFields):]))
}

// ReadInt32 reads the int32 stored in fieldIdx's slot (little-endian).
func (r *RowReader) ReadInt32(fieldIdx int) int32 {
	return int32(binary.LittleEndian.Uint32(r.buf[SlotOffset(fieldIdx, r.numFields):]))
}

// ReadInt64 reads the int64 in fieldIdx's 8-byte slot (little-endian).
func (r *RowReader) ReadInt64(fieldIdx int) int64 {
	return int64(r.readPrimSlot(fieldIdx))
}

// ReadFloat32 reads the IEEE 754 single-precision float stored in fieldIdx's slot.
func (r *RowReader) ReadFloat32(fieldIdx int) float32 {
	return math.Float32frombits(binary.LittleEndian.Uint32(r.buf[SlotOffset(fieldIdx, r.numFields):]))
}

// ReadFloat64 reads the IEEE 754 double-precision float stored in fieldIdx's slot.
func (r *RowReader) ReadFloat64(fieldIdx int) float64 {
	return math.Float64frombits(r.readPrimSlot(fieldIdx))
}

// ReadDate32 reads days-since-Unix-epoch (int32) from fieldIdx's slot.
func (r *RowReader) ReadDate32(fieldIdx int) int32 {
	return int32(binary.LittleEndian.Uint32(r.buf[SlotOffset(fieldIdx, r.numFields):]))
}

// ReadTimestamp reads microseconds-since-Unix-epoch (int64) from fieldIdx's slot.
func (r *RowReader) ReadTimestamp(fieldIdx int) int64 {
	return int64(r.readPrimSlot(fieldIdx))
}

// ReadDuration reads a duration in microseconds (int64) from fieldIdx's slot.
func (r *RowReader) ReadDuration(fieldIdx int) int64 {
	return int64(r.readPrimSlot(fieldIdx))
}

// readVarSlot decodes the (relativeOffset, size) pair from fieldIdx's slot.
func (r *RowReader) readVarSlot(fieldIdx int) (relOffset, size int) {
	slotVal := r.readPrimSlot(fieldIdx)
	return int(slotVal >> 32), int(slotVal & 0xFFFFFFFF)
}

// ReadString reads the variable-length UTF-8 string for fieldIdx (O(1) slot lookup).
func (r *RowReader) ReadString(fieldIdx int) string {
	relOffset, size := r.readVarSlot(fieldIdx)
	return string(r.buf[relOffset : relOffset+size])
}

// ReadBinary reads the raw bytes for fieldIdx (O(1) slot lookup).
// The returned slice shares the underlying buffer; copy if longer lifetime is needed.
func (r *RowReader) ReadBinary(fieldIdx int) []byte {
	relOffset, size := r.readVarSlot(fieldIdx)
	return r.buf[relOffset : relOffset+size]
}

// ReadFixedArray returns an ArrayReader for the fixed-width array stored in fieldIdx.
// elemSize must match the width used when writing (1, 2, 4, or 8).
func (r *RowReader) ReadFixedArray(fieldIdx, elemSize int) *ArrayReader {
	relOffset, size := r.readVarSlot(fieldIdx)
	return NewFixedArrayReader(r.buf[relOffset:relOffset+size], elemSize)
}

// ReadVarArray returns an ArrayReader for the variable-width array (string/binary) stored in fieldIdx.
func (r *RowReader) ReadVarArray(fieldIdx int) *ArrayReader {
	relOffset, size := r.readVarSlot(fieldIdx)
	return NewVarArrayReader(r.buf[relOffset : relOffset+size])
}

// ReadNestedRow returns a RowReader for the nested row stored in fieldIdx.
// numFields must match the field count of the nested schema.
func (r *RowReader) ReadNestedRow(fieldIdx, numFields int) *RowReader {
	relOffset, size := r.readVarSlot(fieldIdx)
	return NewRowReader(r.buf[relOffset:relOffset+size], numFields)
}

// ReadMap returns a MapReader for the map stored in fieldIdx.
func (r *RowReader) ReadMap(fieldIdx int) *MapReader {
	relOffset, size := r.readVarSlot(fieldIdx)
	return NewMapReader(r.buf[relOffset : relOffset+size])
}

File: rust/fory-core/src/row/reader.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use super::{bit_util::calculate_bitmap_width_in_bytes, row::Row};
use byteorder::{ByteOrder, LittleEndian};

struct FieldAccessorHelper<'a> {
    row: &'a [u8],
    get_field_offset: Box<dyn Fn(usize) -> usize>,
}

impl<'a> FieldAccessorHelper<'a> {
    fn read_u32(row: &[u8], offset: usize) -> Option<u32> {
        // in case `row` does not have enough bytes
        let end = offset.checked_add(4)?;
        let bytes = row.get(offset..end)?;
        Some(LittleEndian::read_u32(bytes))
    }

    fn get_offset_size(&self, idx: usize) -> Option<(u32, u32)> {
        let row = self.row;
        let field_offset = (self.get_field_offset)(idx);
        let offset = Self::read_u32(row, field_offset)?;
        let size_offset = field_offset.checked_add(4)?;
        let size = Self::read_u32(row, size_offset)?;
        Some((offset, size))
    }

    pub fn new(
        row: &'a [u8],
        get_field_offset: Box<dyn Fn(usize) -> usize>,
    ) -> FieldAccessorHelper<'a> {
        FieldAccessorHelper {
            row,
            get_field_offset,
        }
    }

    pub fn get_field_bytes(&self, idx: usize) -> &'a [u8] {
        let row = self.row;
        let Some((offset, size)) = self.get_offset_size(idx) else {
            return &[];
        };
        let offset = offset as usize;
        let size = size as usize;
        let Some(end) = offset.checked_add(size) else {
            return &[];
        };
        row.get(offset..end).unwrap_or(&[])
    }
}

pub struct StructViewer<'r> {
    field_accessor_helper: FieldAccessorHelper<'r>,
}

impl<'r> StructViewer<'r> {
    pub fn new(row: &'r [u8], num_fields: usize) -> StructViewer<'r> {
        let bit_map_width_in_bytes = calculate_bitmap_width_in_bytes(num_fields);
        StructViewer {
            field_accessor_helper: FieldAccessorHelper::new(
                row,
                Box::new(move |idx: usize| bit_map_width_in_bytes + idx * 8),
            ),
        }
    }

    pub fn get_field_bytes(&self, idx: usize) -> &'r [u8] {
        self.field_accessor_helper.get_field_bytes(idx)
    }
}

pub struct ArrayViewer<'r> {
    num_elements: usize,
    field_accessor_helper: FieldAccessorHelper<'r>,
}

impl<'r> ArrayViewer<'r> {
    pub fn new(row: &'r [u8]) -> ArrayViewer<'r> {
        let num_elements = row
            .get(0..8)
            .map(|bytes| LittleEndian::read_u64(bytes) as usize)
            .unwrap_or(0);
        let bit_map_width_in_bytes = calculate_bitmap_width_in_bytes(num_elements);
        ArrayViewer {
            num_elements,
            field_accessor_helper: FieldAccessorHelper::new(
                row,
                Box::new(move |idx: usize| 8 + bit_map_width_in_bytes + idx * 8),
            ),
        }
    }

    pub fn num_elements(&self) -> usize {
        self.num_elements
    }

    pub fn get_field_bytes(&self, idx: usize) -> &'r [u8] {
        self.field_accessor_helper.get_field_bytes(idx)
    }
}

pub struct MapViewer<'r> {
    key_row: &'r [u8],
    value_row: &'r [u8],
}

impl<'r> MapViewer<'r> {
    pub fn new(row: &'r [u8]) -> MapViewer<'r> {
        let Some(header) = row.get(0..8) else {
            return MapViewer {
                key_row: &[],
                value_row: &[],
            };
        };
        let key_byte_size = LittleEndian::read_u64(header) as usize;
        let key_end = (8usize).saturating_add(key_byte_size).min(row.len());
        MapViewer {
            value_row: &row[key_end..],
            key_row: &row[8..key_end],
        }
    }

    pub fn get_key_row(&self) -> &[u8] {
        self.key_row
    }

    pub fn get_value_row(&self) -> &[u8] {
        self.value_row
    }
}

pub fn from_row<'a, T: Row<'a>>(row: &'a [u8]) -> T::ReadResult {
    T::cast(row)
}
